SimCalorimetry/HGCalAssociatorProducers/plugins/TSToSimTSHitLCAssociatorByEnergyScoreImpl.cc

```diff
@@ -0,0 +1,259 @@
+    const edm::Handle<ticl::TracksterCollection>& sTCH) const {
+  // Get collections
+  const auto& tracksters = *tCH.product();
+  const auto& layerClusters = *lCCH.product();
+  const auto& sC = *sCCH.product();
+  const auto& cP = *cPCH.product();
+  const auto cPHandle_id = cPCH.id();
+  const auto& simTSs = *sTCH.product();
+  const auto nTracksters = tracksters.size();
+  const auto nSimTracksters = simTSs.size();
+
+  std::unordered_map<DetId, std::vector<std::pair<int, float>>> detIdSimTSId_Map;
+  std::unordered_map<DetId, std::vector<std::pair<int, float>>> detIdSimClusterId_Map;
+  std::unordered_map<DetId, std::vector<std::pair<int, float>>> detIdCaloParticleId_Map;
+  std::unordered_map<DetId, std::vector<std::pair<int, float>>> detIdToRecoTSId_Map;
+
+  hgcal::sharedEnergyAndScore_t recoToSim_sharedEnergyAndScore;
+  hgcal::sharedEnergyAndScore_t simToReco_sharedEnergyAndScore;
+
+  recoToSim_sharedEnergyAndScore.resize(nTracksters);
+  simToReco_sharedEnergyAndScore.resize(nSimTracksters);
+
+  for (size_t i = 0; i < nTracksters; ++i)
+    recoToSim_sharedEnergyAndScore[i].resize(nSimTracksters);
+  for (size_t i = 0; i < nSimTracksters; ++i)
+    simToReco_sharedEnergyAndScore[i].resize(nTracksters);
+
+  // fill sim maps
+
+  for (size_t i = 0; i < sC.size(); ++i) {
+    for (const auto& haf : sC[i].hits_and_fractions()) {
+      detIdSimClusterId_Map[haf.first].emplace_back(i, haf.second);
+    }
+  }
+
+  for (size_t i = 0; i < cP.size(); ++i) {
+    for (const auto& sc : cP[i].simClusters()) {
+      for (const auto& haf : sc->hits_and_fractions()) {
+        auto hitId = haf.first;
+        auto found = std::find_if(detIdCaloParticleId_Map[hitId].begin(),
+                                  detIdCaloParticleId_Map[hitId].end(),
+                                  [=](const std::pair<int, float>& v) { return v.first == static_cast<int>(i); });
+        if (found == detIdCaloParticleId_Map[hitId].end())
+          detIdCaloParticleId_Map[haf.first].emplace_back(i, haf.second);
+        else
+          found->second += haf.second;
+      }
+    }
+  }
+
+  for (size_t i = 0; i < nSimTracksters; ++i) {
+    const auto& lcsInSimTrackster = simTSs[i].vertices();
+    const auto& multiplicities = simTSs[i].vertex_multiplicity();
+    for (size_t j = 0; j < lcsInSimTrackster.size(); ++j) {
+      const auto& v = lcsInSimTrackster[j];
+      float fraction = 1.f / multiplicities[j];
+      for (const auto& haf : layerClusters[v].hitsAndFractions()) {
+        detIdSimTSId_Map[haf.first].emplace_back(i, haf.second * fraction);
+      }
+    }
+  }
+
+  // fill reco map
+
+  for (size_t i = 0; i < nTracksters; ++i) {
+    const auto& lcsInSimTrackster = tracksters[i].vertices();
+    const auto& multiplicities = tracksters[i].vertex_multiplicity();
+    for (size_t j = 0; j < lcsInSimTrackster.size(); ++j) {
+      const auto& v = lcsInSimTrackster[j];
+      float fraction = 1.f / multiplicities[j];
```

Comment:
Is this safe multiplicities[j] = 0 ?

Reply:
Ciao @mandrenguyen, modulo bugs, the multiplicity cannot be 0.

Reply:
I think it's still better to add a check.